ratpack-test/src/main/java/ratpack/test/http/internal/DefaultMultipartForm.java

```diff
@@ -0,0 +1,396 @@
+      }
+    }
+
+    private void file(File file, StringBuilder out) {
+      type(file.getContentType(), out);
+      encoding(file.getEncoding(), out);
+      data(file.getData(), out);
+    }
+
+    private String generateSubBoundary(String boundary, String name) {
+      return String.format("%s_%s", boundary, name);
+    }
+
+    private void type(String value, StringBuilder out) {
+      label("Content-Type", value, out);
+    }
+
+    private void label(String label, String value, StringBuilder out) {
+      out.append(label);
+      out.append(": ");
+      out.append(value);
+      out.append("\n");
+    }
+
+    private void openBoundary(String boundary, StringBuilder out) {
+      out.append("--");
+      out.append(boundary);
+      out.append("\n");
+    }
+
+    private void closeBoundary(String boundary, StringBuilder out) {
+      out.append("--");
+      out.append(boundary);
+      out.append("--\n");
+    }
+
+  }
+
+  final static class Data {
+
+    final String boundary;
+    final Map<String, Set<String>> fields;
+    final Map<String, Set<File>> files;
+
+    Data(String boundary, Map<String, Set<String>> fields, Map<String, Set<File>> files) {
+      this.boundary = boundary;
+      this.fields = Collections.unmodifiableMap(fields);
+      this.files = Collections.unmodifiableMap(files);
+    }
+
+  }
+
+  public final static class Builder implements MultipartFormSpec {
+
+    private final String _boundary;
+    private final Map<String, Set<String>> _fields;
+    private final Map<String, Set<File>> _files;
+
+    private Builder() {
+      this._boundary = generateBoundary();
+      this._fields = new TreeMap<>();
+      this._files = new TreeMap<>();
+    }
+
+    public String getBoundary() {
+      return _boundary;
+    }
+
+    public Map<String, Set<String>> getFields() {
+      return Collections.unmodifiableMap(_fields);
+    }
+
+    public Map<String, Set<File>> getFiles() {
+      return Collections.unmodifiableMap(_files);
+    }
+
+    public DefaultMultipartForm build() {
+      return new DefaultMultipartForm(new Data(_boundary, _fields, _files));
+    }
+
+    public Builder field(String name, String value) {
+      if (!_fields.containsKey(name)) {
+        _fields.put(name, new TreeSet<>());
+      }
+      _fields.get(name).add(value);
+
+      return this;
+    }
+
+    public Builder fields(Map<String, String> data) {
+      data.forEach(this::field);
+
+      return this;
+    }
+
+    public FileBuilder file() {
+      return new FileBuilder(this);
+    }
+
+    private Builder add(File file) {
+      String name = file.getField();
+      if (!_files.containsKey(name)) {
+        _files.put(name, new TreeSet<>());
+      }
+      _files.get(name).add(file);
+
+      return this;
+    }
+
+    private static String generateBoundary() {
+      return String.format("TEST%d", System.currentTimeMillis());
+    }
+
+  }
+
+  public final static class FileBuilder implements MultipartFileSpec {
```

Comment:
Why are there underscores in the names of fields in this class? I don't think such naming convention is used in ratpack's codebase.